jwplatform/upload.py

```diff
@@ -0,0 +1,132 @@
+import os
+from enum import Enum
+from hashlib import md5
+import requests
+from requests import HTTPError
+
+from jwplatform import constants
+
+
+class UploadType(Enum):
+    direct = "direct"
+    multipart = "multipart"
+
+
+def determine_upload_method(file, target_part_size) -> str:
+    filename = file.name
+    file_size = os.stat(filename).st_size
+    if file_size <= target_part_size:
+        return UploadType.direct.value
+    return UploadType.multipart.value
+
+
+class MultipartUpload:
+
+    def __init__(self, client, upload_id: str, file, target_part_size, retry_count):
+        self.upload_id = upload_id
+        self.target_part_size = target_part_size
+        self.upload_retry_count = retry_count
+        self.file = file
+        self.client = client
+
+    def upload(self):
+
+        if self.target_part_size < constants.MIN_PART_SIZE:
+            raise ValueError(f"The part size has to be at least greater than {constants.MIN_PART_SIZE} bytes.")
+
+        filename = self.file.name
+        file_size = os.stat(filename).st_size
+        part_count = file_size // self.target_part_size + 1
+        # Get the part links
+        upload_links = self._get_pre_signed_part_links(part_count)
+
+        # Upload the parts
+        for part_number in range(1, part_count + 1):
+            bytes_chunk = self.file.read(self.target_part_size)
+            if part_number < part_count and len(bytes_chunk) != self.target_part_size:
+                # ToDo: Throw custom exceptions that are handled by the client(user) code.
+                raise IOError("Failed to read enough bytes")
+            retry_count = 0
+            while retry_count < self.upload_retry_count:
```

Comment:
This retry logic is kinda duplicated between the two upload methods, I wonder if there's a good way to make it reusable.

Reply:
The logic to retry is only a few lines (less than 5). Would have ben useful to make it reusable had it been longer.